---
title: "topdownr"
author:
- name: Sebastian Gibb
  affiliation: Department of Anesthesiology and Intensive Care, University Medicine Greifswald, Germany.
- name: Pavel V. Shliaha
  affiliation: Department of Biochemistry and Molecular Biology, University of Southern Denmark, Denmark.
package: topdownr
abstract: >
  This vignette describes the analysis functionality
  of the 'topdownr' package.
output:
  BiocStyle::html_document2:
    toc_float: true
    tidy: TRUE
bibliography: topdownr.bib
vignette: >
  %\VignetteIndexEntry{TopDown Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteKeywords{Mass Spectrometry, Proteomics, Bioinformatics}
  %\VignetteEncoding{UTF-8}
  %\VignettePackage{topdownr}
---

```{r environment, echo=FALSE}
suppressPackageStartupMessages(library("topdownr"))
suppressPackageStartupMessages(library("topdownrdata"))
suppressPackageStartupMessages(library("BiocStyle"))
```

```{r include_forword, child="Foreword.md"}
```

```{r include_bugs, child="Bugs.md"}
```

# Introduction

## The `topdownr` analysis workflow

![topdownr workflow](images/workflow/analysis.mmd.png)

# Working with `topdownr`

```{r load_package}
library("topdownr")
```

## Importing files

The example files are provided in the `topdownrdata` package. For a full
analysis you need a `.fasta` file with the protein sequence, the
`.experiments.csv` files containing the method information, the `.txt` files
containing the scan header information and the `.mzML` files with the
deconvulted spectra.

```{r list_files, eval=-1, echo=1, comment=NA}
list.files(topdownrdata::topDownDataPath("myoglobin"))
lapply(topdownr:::.listTopDownFiles(topdownrdata::topDownDataPath("myoglobin")),
       function(x) { c(head(x, 2), "...") })
```

All these files have to be in a directory. You could import them via
`readTopDownFiles`:

```{r import_files}
library("topdownr")

tds <- readTopDownFiles(topdownrdata::topDownDataPath("myoglobin"))
tds
```

## The `TopDownSet` anatomy

![TopDownSet anatomy, image adopted from [@SummarizedExperiment].](images/TopDownSet.svg)

The `TopDownSet` contains the following components: *Fragment data*, *Condition
data*, *Assay data*.

### Fragment data

```{r rowViews}
rowViews(tds)
```

Fragment data contains one `AAString` (the protein sequence) and an `IRanges`
object that stores the `start`, `end` (and `width`) values of the fragments.
Additionally it has a `DataFrame` for the `mass`, `type` and `z` information
of each fragment.

(*Implementation detail:* the fragment data are represented by an
`FragmentViews` object that is an overloaded `XStringViews` object.)

### Condition data

```{r colData}
conditionData(tds)[, 1:5]
```

Condition data is a `DataFrame` that contains the combined header information
for each MS run (combined from method (`.experiments.csv` files)/scan header
(`.txt` files) table and metadata from the `.mzML` files).

### Assay data

```{r assayData}
assayData(tds)[810:819, 1:10]
```

Assay data is a `sparseMatrix` where the rows correspond to the fragments,
the columns to the runs/conditions and the entries to the intensity values.
A `sparseMatrix` is similar to the classic `matrix` in *R* but stores just
the values that are different from zero.

## Subsetting a `TopDownSet`

A `TopDownSet` could be subsetted by the fragment and the condition data.

```{r subsetting}
# select the first 100 fragments
tds[1:100]

# select all "c" fragments
tds["c"]

# select just the 100. "c" fragment
tds["c100"]

# select all "a" and "b" fragments but just the first 100 "c"
tds[c("a", "b", paste0("c", 1:100))]

# select condition/run 1 to 10
tds[, 1:10]

# select all conditions from one file
tds[, tds$File == "myo_1211_ETDReagentTarget_1e+06_1"]

# select all "c" fragments from a single file
tds["c", tds$File == "myo_1211_ETDReagentTarget_1e+06_1"]
```

***TODO***

```{r sessioninfo}
sessionInfo()
```

# References
