---
title: "topdown"
author:
- name: Sebastian Gibb
  affiliation: Department of Anesthesiology and Intensive Care, University Medicine Greifswald, Germany.
- name: Pavel V. Shliaha
  affiliation: Department of Biochemistry and Molecular Biology, University of Southern Denmark, Denmark.
package: topdown
abstract: >
  This vignette describes the setup and basic functionality
  of the 'topdown' package.
output:
  BiocStyle::html_document2:
    toc_float: true
    tidy: TRUE
bibliography: topdown.bib
vignette: >
  %\VignetteIndexEntry{TopDown}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteKeywords{Mass Spectrometry, Proteomics, Bioinformatics}
  %\VignetteEncoding{UTF-8}
  %\VignettePackage{topdown}
---

```{r environment, echo=FALSE}
suppressPackageStartupMessages(library("topdown"))
suppressPackageStartupMessages(library("topdowndata"))
suppressPackageStartupMessages(library("BiocStyle"))
```

```{r include_forword, child="Foreword.md"}
```

```{r include_bugs, child="Bugs.md"}
```

# Introduction

## The `topdown` workflow

![topdown workflow](images/workflow/workflow.png)

# Setup

## Setup the Thermo Software

1. Install the workstation by running `OrbitrapFusionLumosWorkstation.exe`.
2. Run `C:\Thermo\Instruments\TNG\OrbitrapFusionLumos\2.1\System\programs\TNGInstrumentConfigControl.exe`,
   set *Optional Ion Source* to *ETD and Internal Calibration*, click *Apply* and *OK* (you could ignore the *restart your instrument* message).
3. Replace `C:\Thermo\Instruments\TNG\OrbitrapFusionLumos\2.1\System\programs\Thermo.TNG.Calcium.MethodXML.dll` by `Thermo.TNG.Calcium.MethodXML.dll`.

## Setup XMLMethodChanger

**TODO:**: has to be easier and more user-friendly:
Download and compile https://github.com/thermofisherlsms/meth-modifications

## Setup Operating System

In order to use `XMLMethodChanger` the operating system has to use the `.` (dot)
as decimal mark and the `,` (comma) as digit group separator (one thousand dot
two should be formated as `1,000.2`).

In *Windows 7* the settings are located at `Windows Control Panel > Region and Language > Formats`.
Choose *English (USA)* here or use the *Additional settings* button to change it
manually.

## Setup ScanHeadsman

Download *ScanHeadsman* from https://bitbucket.org/caetera/scanheadsman

It requires Microsoft **.NET 4.5** or later (it is often preinstalled on a typical modern Windows
or could be found in Microsoft's Download Center, e.g.
https://www.microsoft.com/en-us/download/details.aspx?id=30653).
Additionally you would need Thermo's *MS File Reader* which could be downloaded
free of charge (but you have to register) from the Thermo FlexNet website:
https://thermo.flexnetoperations.com/

*ScanHeadsman* was created by Vladimir Gorshkov.

# Working with `topdown`

```{r load_package}
library("topdown")
```

## Importing files

The example files are provided in the `topdowndata` package. For a full
analysis you need a `.fasta` file with the protein sequence, the
`.experiments.csv` files containing the method information, the `.txt` files
containing the scan header information and the `.mzML` files with the
deconvulted spectra.

```{r list_files, eval=-1, echo=1, comment=NA}
list.files(topdowndata::topDownDataPath("myoglobin"))
lapply(topdown:::.listTopDownFiles(topdowndata::topDownDataPath("myoglobin")),
       function(x) { c(head(x, 2), "...") })
```

All these files have to be in a directory. You could import them via
`readTopDownFiles`:

```{r import_files}
library("topdown")

tds <- readTopDownFiles(topdowndata::topDownDataPath("myoglobin"))
tds
```

## The `TopDownSet` anatomy

![TopDownSet anatomy, image adopted from [@SummarizedExperiment].](images/TopDownSet.svg)

The `TopDownSet` contains the following components: *Fragment data*, *Condition
data*, *Assay data*.

### Fragment data

```{r rowViews}
fragmentData(tds)
```

Fragment data contains one `AAString` (the protein sequence) and an `IRanges`
object that stores the `start`, `end` (and `width`) values of the fragments.
Additionally it has a `DataFrame` for the `mass`, `type` and `z` information
of each fragment.

(*Implementation detail:* the fragment data are represented by an
`FragmentViews` object that is an overloaded `XStringViews` object.)

### Condition data

```{r colData}
conditionData(tds)[, 1:5]
```

Condition data is a `DataFrame` that contains the combined header information
for each MS run (combined from method (`.experiments.csv` files)/scan header
(`.txt` files) table and metadata from the `.mzML` files).

### Assay data

```{r assayData}
assayData(tds)[810:819, 1:10]
```

Assay data is a `sparseMatrix` where the rows correspond to the fragments,
the columns to the runs/conditions and the entries to the intensity values.
A `sparseMatrix` is similar to the classic `matrix` in *R* but stores just
the values that are different from zero.

## Subsetting a `TopDownSet`

A `TopDownSet` could be subsetted by the fragment and the condition data.

```{r subsetting}
# select the first 100 fragments
tds[1:100]

# select all "c" fragments
tds["c"]

# select just the 100. "c" fragment
tds["c100"]

# select all "a" and "b" fragments but just the first 100 "c"
tds[c("a", "b", paste0("c", 1:100))]

# select condition/run 1 to 10
tds[, 1:10]

# select all conditions from one file
tds[, tds$File == "myo_1211_ETDReagentTarget_1e+06_1"]

# select all "c" fragments from a single file
tds["c", tds$File == "myo_1211_ETDReagentTarget_1e+06_1"]
```

```{r sessioninfo}
sessionInfo()
```

# References
