<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fragmentation Analysis with topdownr • topdownr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">topdownr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/analysis.html">Fragmentation Analysis with topdownr</a>
    </li>
    <li>
      <a href="../articles/data-generation.html">Data Generation for topdownr</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sgibb/topdownr/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Fragmentation Analysis with topdownr</h1>
                        <h4 class="author">Pavel V. Shliaha</h4>
            <address class="author_afil">
      Department of Biochemistry and Molecular Biology, University of Southern Denmark, Denmark.<br><h4 class="author">Sebastian Gibb</h4>
            <address class="author_afil">
      Department of Anesthesiology and Intensive Care, University Medicine Greifswald, Germany.<br><h4 class="author">Ole Nørregaard Jensen</h4>
            <address class="author_afil">
      Department of Biochemistry and Molecular Biology, University of Southern Denmark, Denmark.<br>
</address>
</address>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette describes the fragmentation analysis functionality of the <code>topdownr</code> package.</p>
    </div>
    
<div class="contents">
<div id="foreword" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#foreword" class="anchor"></a>Foreword</h1>
<p><a href="http://bioconductor.org/packages/topdownr"><em>topdownr</em></a> is free and open-source software.</p>
</div>
<div id="questions-and-bugs" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#questions-and-bugs" class="anchor"></a>Questions and bugs</h1>
<p>For bugs, typos, suggestions or other questions, please file an issue in our tracking system (<a href="https://github.com/sgibb/topdownr/issues" class="uri">https://github.com/sgibb/topdownr/issues</a>) providing as much information as possible, a reproducible example and the output of <code>sessionInfo()</code>.</p>
<p>If you don’t have a GitHub account or wish to reach a broader audience for general questions about proteomics analysis using R, you may want to use the Bioconductor support site: <a href="https://support.bioconductor.org/" class="uri">https://support.bioconductor.org/</a>.</p>
</div>
<div id="introductionworking-with-topdownr" class="section level1">
<h1 class="hasAnchor">
<a href="#introductionworking-with-topdownr" class="anchor"></a>Introduction/Working with <code>topdownr</code>
</h1>
<p>Load the package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"topdownr"</span>)</code></pre></div>
<div id="importing-files" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-files" class="anchor"></a>Importing Files</h2>
<p>Some example files are provided in the <code>topdownrdata</code> package. For a full analysis you need a <code>.fasta</code> file with the protein sequence, the <code>.experiments.csv</code> files containing the method information, the <code>.txt</code> files containing the scan header information and the <code>.mzML</code> files with the deconvulted spectra.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## list.files(topdownrdata::topDownDataPath("myoglobin"))</code></pre></div>
<pre><code>$csv
[1] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/experimentsmyo_1211_ETDReagentTarget_1e+06_1.experiments.csv.gz2"
[2] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/experimentsmyo_1211_ETDReagentTarget_1e+06_2.experiments.csv.gz2"
[3] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/experimentsmyo_1211_ETDReagentTarget_1e+07_1.experiments.csv.gz2"
[4] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/experimentsmyo_1211_ETDReagentTarget_1e+07_2.experiments.csv.gz2"
[5] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/experimentsmyo_1211_ETDReagentTarget_5e+06_1.experiments.csv.gz2"
[6] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/experimentsmyo_1211_ETDReagentTarget_5e+06_2.experiments.csv.gz2"
[7] "..."                                                                                                                                               

$fasta
[1] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/fastamyoglobin.fasta.gz2"
[2] "..."                                                                                                       

$mzML
[1] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/mzmlmyo_1211_ETDReagentTarget_1e+06_1.mzML.gz2"
[2] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/mzmlmyo_1211_ETDReagentTarget_1e+06_2.mzML.gz2"
[3] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/mzmlmyo_1211_ETDReagentTarget_1e+07_1.mzML.gz2"
[4] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/mzmlmyo_1211_ETDReagentTarget_1e+07_2.mzML.gz2"
[5] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/mzmlmyo_1211_ETDReagentTarget_5e+06_1.mzML.gz2"
[6] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/mzmlmyo_1211_ETDReagentTarget_5e+06_2.mzML.gz2"
[7] "..."                                                                                                                             

$txt
[1] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/headermyo_1211_ETDReagentTarget_1e+06_1.txt.gz2"
[2] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/headermyo_1211_ETDReagentTarget_1e+06_2.txt.gz2"
[3] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/headermyo_1211_ETDReagentTarget_1e+07_1.txt.gz2"
[4] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/headermyo_1211_ETDReagentTarget_1e+07_2.txt.gz2"
[5] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/headermyo_1211_ETDReagentTarget_5e+06_1.txt.gz2"
[6] "...//home/sebastian/R/x86_64-pc-linux-gnu-library/3.3/topdownrdata/2017_06_29_myo/headermyo_1211_ETDReagentTarget_5e+06_2.txt.gz2"
[7] "..."                                                                                                                              </code></pre>
<p>All these files have to be in a directory. You could import them via <code>readTopDownFiles</code>. This function has some arguments. The most important ones are the <code>path</code> of the directory containing the files, the protein <code>modification</code> (e.g. initiator methionine removal, <code>"Met-loss"</code>), and adducts (e.g. proton transfer often occurs from c to z-fragment after ETD reaction).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## the mass adduct for a proton
H &lt;-<span class="st"> </span><span class="fl">1.0078250321</span>

myoglobin &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readTopDownFiles.html">readTopDownFiles</a></span>(
    ## directory path
    <span class="dt">path =</span> topdownrdata::<span class="kw">topDownDataPath</span>(<span class="st">"myoglobin"</span>),
    ## fragmentation types
    <span class="dt">type =</span> <span class="kw">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>, <span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span>),
    ## adducts (add -H/H to c/z and name
    ## them cmH/zpH (c minus H, z plus H)
    <span class="dt">adducts =</span> <span class="kw">data.frame</span>(
        <span class="dt">mass=</span><span class="kw">c</span>(-H, H),
        <span class="dt">to=</span><span class="kw">c</span>(<span class="st">"c"</span>, <span class="st">"z"</span>),
        <span class="dt">name=</span><span class="kw">c</span>(<span class="st">"cmH"</span>, <span class="st">"zpH"</span>)),
    ## initiator methionine removal
    <span class="dt">modifications =</span> <span class="st">"Met-loss"</span>,
    ## don't use neutral loss
    <span class="dt">neutralLoss =</span> <span class="ot">NULL</span>,
    ## tolerance for fragment matching
    <span class="dt">tolerance =</span> <span class="fl">5e-6</span>
)</code></pre></div>
<pre><code>## Warning in FUN(X[[i]], ...): 61 FilterString entries modified because of
## duplicated ID for different conditions.</code></pre>
<pre><code>## Warning in FUN(X[[i]], ...): 63 FilterString entries modified because of
## duplicated ID for different conditions.</code></pre>
<pre><code>## Warning in FUN(X[[i]], ...): 53 FilterString entries modified because of
## duplicated ID for different conditions.</code></pre>
<pre><code>## Warning in FUN(X[[i]], ...): 55 FilterString entries modified because of
## duplicated ID for different conditions.</code></pre>
<pre><code>## Warning in FUN(X[[i]], ...): 50 FilterString entries modified because of
## duplicated ID for different conditions.

## Warning in FUN(X[[i]], ...): 50 FilterString entries modified because of
## duplicated ID for different conditions.</code></pre>
<pre><code>## Warning in FUN(X[[i]], ...): ID in FilterString are not sorted in ascending
## order. Introduce own condition ID via 'cumsum'.

## Warning in FUN(X[[i]], ...): ID in FilterString are not sorted in ascending
## order. Introduce own condition ID via 'cumsum'.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myoglobin</code></pre></div>
<pre><code>## TopDownSet object (7.01 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1216 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;16910.93]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 5882 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1216x5882 (5.21% != 0)
## Number of matched fragments: 372991 
## Intensity range: [87.61;10704001.00]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).</code></pre>
</div>
<div id="the-topdownset-anatomy" class="section level2">
<h2 class="hasAnchor">
<a href="#the-topdownset-anatomy" class="anchor"></a>The <code>TopDownSet</code> Anatomy</h2>
<p>The assembled object is an <code>TopDownSet</code> object. Briefly it is composed of three interconnected tables:</p>
<ol style="list-style-type: decimal">
<li>
<code>rowViews</code>/<em>fragment data</em>: holds the information on the type of fragments, their modifications and adducts.</li>
<li>
<code>colData</code>/<em>condition data</em>: contains the corresponding fragmentation condition for every spectrum.</li>
<li>
<code>assayData</code>: contains the intensity of assigned fragments.</li>
</ol>
<div class="figure">
<img src="images/TopDownSet.svg" alt="TopDownSet anatomy, image adopted from (Morgan et al. 2017)."><p class="caption">TopDownSet anatomy, image adopted from <span class="citation">(Morgan et al. 2017)</span>.</p>
</div>
</div>
<div id="technical-details" class="section level2">
<h2 class="hasAnchor">
<a href="#technical-details" class="anchor"></a>Technical Details</h2>
<p>This section explains the implementation details of the <code>TopDownSet</code> class. It is not necessary to understand everything written here to use <code>topdownr</code> for the analysis of fragmentation data.</p>
<p>The <code>TopDownSet</code> contains the following components: <em>Fragment data</em>, <em>Condition data</em>, <em>Assay data</em>.</p>
<div id="fragment-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fragment-data" class="anchor"></a>Fragment data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/AbstractTopDownSet-class.html">rowViews</a></span>(myoglobin)</code></pre></div>
<pre><code>## FragmentViews on a 153-letter sequence:
##   GLSDGEWQQVLNVWGKVEADIAGHGQEVLIRLFTG...HPGDFGADAQGAMTKALELFRNDIAAKYKELGFQG
## Mass:
##   16922.95406
## Modifications:
##   Met-loss
## Views:
##        start end width     mass name   type   z                            
##    [1]     1   1     1    30.03 a1     a      1 [G]                        
##    [2]     1   1     1    58.03 b1     b      1 [G]                        
##    [3]     1   1     1    59.01 z1     z      1 [G]                        
##    [4]     1   1     1    60.02 zpH1   z      1 [G]                        
##    [5]     1   1     1    74.05 cmH1   c      1 [G]                        
##    ...   ... ...   ...      ... ...    ...  ... ...                        
## [1212]     2 153   152 16868.93 zpH152 z      1 [LSDGEWQQVLN...AAKYKELGFQG]
## [1213]     1 152   152 16882.96 cmH152 c      1 [GLSDGEWQQVL...IAAKYKELGFQ]
## [1214]     1 152   152 16883.97 c152   c      1 [GLSDGEWQQVL...IAAKYKELGFQ]
## [1215]     2 153   152 16884.95 y152   y      1 [LSDGEWQQVLN...AAKYKELGFQG]
## [1216]     2 153   152 16910.93 x152   x      1 [LSDGEWQQVLN...AAKYKELGFQG]</code></pre>
<p>The fragmentation data are represented by an <code>FragmentViews</code> object that is an overloaded <code>XStringViews</code> object. It contains one <code>AAString</code> (the protein sequence) and an <code>IRanges</code> object that stores the <code>start</code>, <code>end</code> (and <code>width</code>) values of the fragments. Additionally it has a <code>DataFrame</code> for the <code>mass</code>, <code>type</code> and <code>z</code> information of each fragment.</p>
</div>
<div id="condition-data" class="section level3">
<h3 class="hasAnchor">
<a href="#condition-data" class="anchor"></a>Condition data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/AbstractTopDownSet-class.html">conditionData</a></span>(myoglobin)[, <span class="dv">1</span>:<span class="dv">5</span>]</code></pre></div>
<pre><code>## DataFrame with 5882 rows and 5 columns
##                                                                      File
##                                                                     &lt;Rle&gt;
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_01  myo_707_ETDReagentTarget_1e+06_1
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_02  myo_707_ETDReagentTarget_1e+06_1
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_03  myo_707_ETDReagentTarget_1e+06_2
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_04  myo_707_ETDReagentTarget_1e+06_2
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_05  myo_707_ETDReagentTarget_1e+07_1
## ...                                                                   ...
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_1  myo_1211_ETDReagentTarget_1e+07_1
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_2  myo_1211_ETDReagentTarget_1e+07_2
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_3  myo_1211_ETDReagentTarget_1e+07_2
## C1211.70_1.0e+06_1.0e+07_50.00_35_00_1  myo_1211_ETDReagentTarget_1e+07_1
## C1211.70_1.0e+06_1.0e+07_50.00_35_00_2  myo_1211_ETDReagentTarget_1e+07_2
##                                          Scan PeaksCount TotIonCurrent
##                                         &lt;Rle&gt;      &lt;Rle&gt;     &lt;numeric&gt;
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_01    13         64      16904906
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_02    14         69      17520226
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_03    13        109      19653865
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_04    14         85      20301836
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_05    13         75      17174796
## ...                                       ...        ...           ...
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_1    293        306       4618113
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_2    294        188       2603466
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_3    295        150       2104302
## C1211.70_1.0e+06_1.0e+07_50.00_35_00_1    295        256       3725114
## C1211.70_1.0e+06_1.0e+07_50.00_35_00_2    297        135       1786274
##                                         RetentionTime
##                                             &lt;numeric&gt;
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_01      65.69495
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_02      80.23453
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_03      65.57049
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_04      80.12018
## C0707.30_1.0e+05_0.0e+00_00.00_00_07_05      65.56137
## ...                                               ...
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_1       6778.102
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_2       6751.959
## C1211.70_1.0e+06_1.0e+07_50.00_28_00_3       6779.711
## C1211.70_1.0e+06_1.0e+07_50.00_35_00_1       6834.398
## C1211.70_1.0e+06_1.0e+07_50.00_35_00_2       6835.891</code></pre>
<p>Condition data is a <code>DataFrame</code> that contains the combined header information for each MS run (combined from method (<code>.experiments.csv</code> files)/scan header (<code>.txt</code> files) table and metadata from the <code>.mzML</code> files).</p>
</div>
<div id="assay-data" class="section level3">
<h3 class="hasAnchor">
<a href="#assay-data" class="anchor"></a>Assay data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assayData</span>(myoglobin)[<span class="dv">206</span>:<span class="dv">215</span>, <span class="dv">1</span>:<span class="dv">10</span>]</code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class "dgCMatrix"</code></pre>
<pre><code>##    [[ suppressing 10 column names 'C0707.30_1.0e+05_0.0e+00_00.00_00_07_01', 'C0707.30_1.0e+05_0.0e+00_00.00_00_07_02', 'C0707.30_1.0e+05_0.0e+00_00.00_00_07_03' ... ]]</code></pre>
<pre><code>##                                                                          
## z26        .         .         .         .         .   124981.9      .   
## zpH26      .         .         .         .         .        .        .   
## y26   508116.78 454086.72 483405.88 428070.62 374857.8 453643.2 367292.47
## b27        .         .         .         .         .        .        .   
## cmH27      .         .         .         .         .        .        .   
## c27        .         .         .         .         .        .        .   
## x26        .         .         .         .         .        .        .   
## z27    18677.47  24941.96  42561.77  36388.28      .        .    24822.07
## zpH27      .         .         .         .         .        .        .   
## y27   366331.19 377103.00 309914.03 322329.38 323019.6 335588.8 277860.31
##                                  
## z26        .         .        .  
## zpH26      .         .        .  
## y26   389549.88 354230.6 340419.0
## b27        .         .        .  
## cmH27      .         .        .  
## c27        .         .        .  
## x26        .         .        .  
## z27    11467.17      .        .  
## zpH27      .         .        .  
## y27   305515.78 287764.5 271953.4</code></pre>
<p>Assay data is a <code>sparseMatrix</code> from the <code>Matrix</code> package (in detail a <code>dgCMatrix</code>) where the rows correspond to the fragments, the columns to the runs/conditions and the entries to the intensity values. A <code>sparseMatrix</code> is similar to the classic <code>matrix</code> in <em>R</em> but stores just the values that are different from zero.</p>
</div>
</div>
<div id="subsetting-a-topdownset" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-a-topdownset" class="anchor"></a>Subsetting a <code>TopDownSet</code>
</h2>
<p>A <code>TopDownSet</code> could be subsetted by the fragment and the condition data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select the first 100 fragments</span>
myoglobin[<span class="dv">1</span>:<span class="dv">100</span>]</code></pre></div>
<pre><code>## TopDownSet object (3.25 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 100 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;1426.70]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 5882 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 100x5882 (9.68% != 0)
## Number of matched fragments: 56956 
## Intensity range: [105.70;1076768.00]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:45] Subsetted 372991 fragments [1216;5882] to 56956 fragments [100;5882].</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select all "c" fragments</span>
myoglobin[<span class="st">"c"</span>]</code></pre></div>
<pre><code>## TopDownSet object (4.22 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 304 
## Theoretical fragment types (1): c
## Theoretical mass range: [74.05;16883.97]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 5882 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 304x5882 (7.80% != 0)
## Number of matched fragments: 139458 
## Intensity range: [87.61;1203763.75]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 139458 fragments [304;5882].</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select just the 100. "c" fragment</span>
myoglobin[<span class="st">"c100"</span>]</code></pre></div>
<pre><code>## TopDownSet object (2.58 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1 
## Theoretical fragment types (1): c
## Theoretical mass range: [11085.96;11085.96]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 5882 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1x5882 (0.09% != 0)
## Number of matched fragments: 5 
## Intensity range: [1276.91;17056.12]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 5 fragments [1;5882].</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select all "a" and "b" fragments but just the first 100 "c"</span>
myoglobin[<span class="kw">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>, <span class="kw">paste0</span>(<span class="st">"c"</span>, <span class="dv">1</span>:<span class="dv">100</span>))]</code></pre></div>
<pre><code>## TopDownSet object (4.30 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 404 
## Theoretical fragment types (3): a, b, c
## Theoretical mass range: [30.03;16866.94]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 5882 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 404x5882 (6.12% != 0)
## Number of matched fragments: 145348 
## Intensity range: [87.61;1630533.12]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 145348 fragments [404;5882].</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select condition/run 1 to 10</span>
myoglobin[, <span class="dv">1</span>:<span class="dv">10</span>]</code></pre></div>
<pre><code>## TopDownSet object (0.23 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1216 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;16910.93]
## - - - Condition data - - -
## Number of conditions: 1 
## Number of scans: 10 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1216x10 (2.23% != 0)
## Number of matched fragments: 271 
## Intensity range: [7996.26;5499596.00]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 271 fragments [1216;10].</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select all conditions from one file</span>
myoglobin[, myoglobin$File ==<span class="st"> "myo_1211_ETDReagentTarget_1e+06_1"</span>]</code></pre></div>
<pre><code>## TopDownSet object (0.67 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1216 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;16910.93]
## - - - Condition data - - -
## Number of conditions: 222 
## Number of scans: 351 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1216x351 (4.66% != 0)
## Number of matched fragments: 19889 
## Intensity range: [126.42;1454253.25]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 19889 fragments [1216;351].</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select all "c" fragments from a single file</span>
myoglobin[<span class="st">"c"</span>, myoglobin$File ==<span class="st"> "myo_1211_ETDReagentTarget_1e+06_1"</span>]</code></pre></div>
<pre><code>## TopDownSet object (0.40 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 304 
## Theoretical fragment types (1): c
## Theoretical mass range: [74.05;16883.97]
## - - - Condition data - - -
## Number of conditions: 222 
## Number of scans: 351 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 304x351 (6.05% != 0)
## Number of matched fragments: 6455 
## Intensity range: [251.15;1025625.38]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 6455 fragments [304;351].</code></pre>
</div>
</div>
<div id="fragmentation-data-analysis-of-myoglobin" class="section level1">
<h1 class="hasAnchor">
<a href="#fragmentation-data-analysis-of-myoglobin" class="anchor"></a>Fragmentation Data Analysis of Myoglobin</h1>
<p>We follow the following workflow:</p>
<div class="figure">
<img src="images/workflow/analysis.mmd.png" alt="topdownr workflow"><p class="caption">topdownr workflow</p>
</div>
<p>We use the example data loaded in <a href="analysis.html#21_importing_files">Importing Files</a>.</p>
<p>The data contains several replicates for each fragmentation condition. Before aggregation can be performed we need to remove scans with inadequate injection times and fragments with low intensity or poor intensity reproducibility.</p>
<div id="filter-conditions-on-injection-times" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-conditions-on-injection-times" class="anchor"></a>Filter Conditions on Injection Times</h2>
<p>Injection times should be consistent for a particular <em>m/z</em> and particular AGC target. High or low injection times indicate problems with on-the-flight AGC calculation or spray instability for a particular scan. Hence the <code>topdownr</code> automatically calculates median injection time for a given <em>m/z</em> and AGC target combination. The user can choose to remove all scans that deviate more than a certain amount from the corresponding median and/or choose to keep <code>N</code> scans with the lowest deviation from the median for every condition.</p>
<p>Here we show an example of such filtering and the effect on the distribution of injection times.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">injTimeBefore &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AbstractTopDownSet-class.html">colData</a></span>(myoglobin)
injTimeBefore$Status &lt;-<span class="st"> "before filtering"</span>

## filtering on max deviation and just keep the
## 2 technical replicates per condition with the
## lowest deviation
myoglobin &lt;-<span class="st"> </span><span class="kw"><a href="../reference/TopDownSet-class.html">filterInjectionTime</a></span>(
    myoglobin,
    <span class="dt">maxDeviation =</span> <span class="kw">log2</span>(<span class="dv">3</span>),
    <span class="dt">keepTopN =</span> <span class="dv">2</span>
)

myoglobin</code></pre></div>
<pre><code>## TopDownSet object (4.89 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1216 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;16910.93]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 3696 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1216x3696 (5.70% != 0)
## Number of matched fragments: 256215 
## Intensity range: [109.29;8493567.00]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 256215 fragments [1216;3696].
## [2017-10-04 22:11:46] 2186 scans filtered with injection time deviation &gt;= 1.58496250072116 or rank &gt;= 3; 256215 fragments [1216;3696].</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">injTimeAfter &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AbstractTopDownSet-class.html">colData</a></span>(myoglobin)
injTimeAfter$Status &lt;-<span class="st"> "after filtering"</span>

injTime &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">rbind</span>(injTimeBefore, injTimeAfter))

## use ggplot for visualisation
<span class="kw">library</span>(<span class="st">"ggplot2"</span>)

<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(injTime,
       <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> <span class="kw">as.factor</span>(AgcTarget),
           <span class="dt">y =</span> IonInjectionTimeMs,
           <span class="dt">group =</span> AgcTarget)) +
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_boxplot">geom_boxplot</a></span>() +
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid">facet_grid</a></span>(Status ~<span class="st"> </span>Mz)</code></pre></div>
<p><img src="analysis_files/figure-html/filterInjectionTimes-1.png" width="672"></p>
</div>
<div id="filter-fragments-on-cv" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-fragments-on-cv" class="anchor"></a>Filter Fragments on CV</h2>
<p>High CV of intensity for a fragment suggests either fragment contamination by another <em>m/z</em> species or problems with deisotoping and we recommend removing all fragments with CV &gt; 30, as shown below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myoglobin &lt;-<span class="st"> </span><span class="kw"><a href="../reference/TopDownSet-class.html">filterCv</a></span>(myoglobin, <span class="dt">threshold=</span><span class="dv">30</span>)
myoglobin</code></pre></div>
<pre><code>## TopDownSet object (4.43 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1216 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;16910.93]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 3696 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1216x3696 (4.82% != 0)
## Number of matched fragments: 216651 
## Intensity range: [109.29;8493567.00]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 256215 fragments [1216;3696].
## [2017-10-04 22:11:46] 2186 scans filtered with injection time deviation &gt;= 1.58496250072116 or rank &gt;= 3; 256215 fragments [1216;3696].
## [2017-10-04 22:11:48] 39564 fragments with CV &gt; 30% filtered; 216651 fragments [1216;3696].</code></pre>
</div>
<div id="filter-fragments-on-intensity" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-fragments-on-intensity" class="anchor"></a>Filter Fragments on Intensity</h2>
<p>When optimizing protein fragmentation we also want to focus on the most intense fragments, hence we recommend removing all low intensity fragments from analysis.</p>
<p>Low intensity is defined relatively to the most intense observation for this fragment (i.e. relatively to the maximum value in an <code>assayData</code> row). In the example below all intensity values, which have less than 10% intensity of the highest intensity to their corresponding fragment (in their corresponding row) are removed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myoglobin &lt;-<span class="st"> </span><span class="kw"><a href="../reference/TopDownSet-class.html">filterIntensity</a></span>(myoglobin, <span class="dt">threshold=</span><span class="fl">0.1</span>)
myoglobin</code></pre></div>
<pre><code>## TopDownSet object (3.57 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1216 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;16910.93]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 3696 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1216x3696 (3.14% != 0)
## Number of matched fragments: 141020 
## Intensity range: [219.52;8493567.00]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 256215 fragments [1216;3696].
## [2017-10-04 22:11:46] 2186 scans filtered with injection time deviation &gt;= 1.58496250072116 or rank &gt;= 3; 256215 fragments [1216;3696].
## [2017-10-04 22:11:48] 39564 fragments with CV &gt; 30% filtered; 216651 fragments [1216;3696].
## [2017-10-04 22:11:48] 75631 intensity values &lt; 0.1 (relative) filtered; 141020 fragments [1216;3696].</code></pre>
</div>
<div id="data-aggregation" class="section level2">
<h2 class="hasAnchor">
<a href="#data-aggregation" class="anchor"></a>Data Aggregation</h2>
<p>The next step of analysis is aggregating technical replicates of fragmentation conditions (columns of <code>assayData</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myoglobin &lt;-<span class="st"> </span><span class="kw">aggregate</span>(myoglobin)
myoglobin</code></pre></div>
<pre><code>## TopDownSet object (2.13 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## Mass : 16922.95
## Modifications (1): Met-loss
## - - - Fragment data - - -
## Number of theoretical fragments: 1216 
## Theoretical fragment types (6): a, b, c, x, y, z
## Theoretical mass range: [30.03;16910.93]
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 1852 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Intensity data - - -
## Size of array: 1216x1852 (3.95% != 0)
## Number of matched fragments: 89063 
## Intensity range: [219.52;8492743.50]
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 256215 fragments [1216;3696].
## [2017-10-04 22:11:46] 2186 scans filtered with injection time deviation &gt;= 1.58496250072116 or rank &gt;= 3; 256215 fragments [1216;3696].
## [2017-10-04 22:11:48] 39564 fragments with CV &gt; 30% filtered; 216651 fragments [1216;3696].
## [2017-10-04 22:11:48] 75631 intensity values &lt; 0.1 (relative) filtered; 141020 fragments [1216;3696].
## [2017-10-04 22:11:48] Aggregated 141020 fragments [1216;3696] to 89063 fragments [1216;1852].</code></pre>
</div>
<div id="random-forest" class="section level2">
<h2 class="hasAnchor">
<a href="#random-forest" class="anchor"></a>Random Forest</h2>
<p>To examine which of the features (fragmentation parameters) have the highest overall impact for a protein we perform random forest machine learning using the <code>ranger</code> <span class="citation">(Wright and Ziegler 2017)</span> <code>R</code>-package.</p>
<p>Before we compute some fragmentation statistics (number of assigned fragments, total assigned intensity, etc.).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"ranger"</span>)

## statistics
<span class="kw">head</span>(<span class="kw">summary</span>(myoglobin))</code></pre></div>
<pre><code>##   Fragments     Total      Min       Q1    Median      Mean       Q3
## 1        27 9728506.2 16043.55 36608.61 123825.95 8000.4163 324536.7
## 2        30 7529430.7 21894.47 71422.85 103939.98 6191.9661 215970.9
## 3         9  756978.5 40572.99 49291.46  60046.72  622.5152 121674.7
## 4         3  335156.1 55048.98 90398.54 125748.09  275.6218 140053.6
## 5         2  211671.4 75538.09 90686.89 105835.69  174.0719 120984.5
## 6         0       0.0     0.00     0.00      0.00    0.0000      0.0
##         Max
## 1 4371568.4
## 2 2545802.2
## 3  158469.6
## 4  154359.0
## 5  136133.3
## 6       0.0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## number of fragments
nFragments &lt;-<span class="st"> </span><span class="kw">summary</span>(myoglobin)$Fragments

## features of interest
foi &lt;-<span class="st"> </span><span class="kw">c</span>(
    <span class="st">"AgcTarget"</span>,
    <span class="st">"EtdReagentTarget"</span>,
    <span class="st">"EtdActivation"</span>,
    <span class="st">"CidActivation"</span>,
    <span class="st">"HcdActivation"</span>,
    <span class="st">"Charge"</span>
)

rfTable &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">cbind</span>(
    <span class="kw">scale</span>(<span class="kw">as.data.frame</span>(<span class="kw"><a href="../reference/AbstractTopDownSet-class.html">colData</a></span>(myoglobin)[foi])),
    <span class="dt">Fragments =</span> nFragments
))

featureImportance &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ranger/topics/ranger">ranger</a></span>(
    Fragments ~<span class="st"> </span>.,
    <span class="dt">data =</span> rfTable,
    <span class="dt">importance =</span> <span class="st">"impurity"</span>
)$variable.importance

<span class="kw">barplot</span>(
    featureImportance/<span class="kw">sum</span>(featureImportance),
    <span class="dt">cex.names =</span> <span class="fl">0.7</span>
)</code></pre></div>
<p><img src="analysis_files/figure-html/randomForest-1.png" width="672"></p>
<p>The two parameters having the lowest overall impact in the <code>myoglobin</code> dataset across all conditions are ETD reagent Target (<code>EtdReagentTarget</code>) and CID activation energy (<code>CidActivation</code>), while ETD reaction time (<code>EtdReactionTime</code>) demonstrates the highest overall impact.</p>
</div>
<div id="combining-fragmentation-conditions-to-maximize-coverage" class="section level2">
<h2 class="hasAnchor">
<a href="#combining-fragmentation-conditions-to-maximize-coverage" class="anchor"></a>Combining Fragmentation Conditions to Maximize Coverage</h2>
<p>The purpose of <code>topdownr</code> is to investigate how maximum coverage with high intensity fragments can be achieved with minimal instrument time. Therefore <code>topdownr</code> reports the best combination of fragmentation conditions (with user specified number of conditions) that covers the highest number of different bonds.</p>
<p>Different fragmentation methods predominantly generate different types of fragments (e.g. b and y for HCD and CID, c and z for ETD, a and x for UVPD).</p>
<p>However N-terminal (a, b and c) as well as C-terminal (x, y and z) fragments originating from the same bond, cover the same number of amino acid sidechains. Hence different types of N-terminal (a, b and c) or C-terminal (x, y and z) fragments from the same bond add no extra sequence information.</p>
<p>Before we compute combinations all the fragments are converted to either N- or C-terminal, as shown in the image below.</p>
<div class="figure">
<img src="images/ncb.svg" alt="Schema of N-/C-terminal fragments or both."><p class="caption">Schema of N-/C-terminal fragments or both.</p>
</div>
<p>In <code>topdownr</code> we convert the <code>TopDownSet</code> into an <code>NCBSet</code> object (<em>N</em>-terminal/<em>C</em>-terminal/<em>B</em>oth).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myoglobinNcb &lt;-<span class="st"> </span><span class="kw">as</span>(myoglobin, <span class="st">"NCBSet"</span>)
myoglobinNcb</code></pre></div>
<pre><code>## NCBSet object (1.94 Mb)
## - - - Protein data - - -
## Amino acid sequence (153): GLSDGEWQQVLNVWGKVEADIAG...MTKALELFRNDIAAKYKELGFQG 
## - - - Fragment data - - -
## Number of N-terminal fragments: 30569
## Number of C-terminal fragments: 29579
## Number of N- and C-terminal fragments: 9306
## - - - Condition data - - -
## Number of conditions: 1852 
## Number of scans: 1852 
## Condition variables (58): File, Scan, ..., Sample, Charge
## - - - Assay data - - -
## Size of array: 152x1852 (24.67% != 0)
## - - - Processing information - - -
## [2017-10-04 22:11:45] 372991 fragments [1216;5882] matched (tolerance: 5 ppm).
## [2017-10-04 22:11:46] Subsetted 372991 fragments [1216;5882] to 256215 fragments [1216;3696].
## [2017-10-04 22:11:46] 2186 scans filtered with injection time deviation &gt;= 1.58496250072116 or rank &gt;= 3; 256215 fragments [1216;3696].
## [2017-10-04 22:11:48] 39564 fragments with CV &gt; 30% filtered; 216651 fragments [1216;3696].
## [2017-10-04 22:11:48] 75631 intensity values &lt; 0.1 (relative) filtered; 141020 fragments [1216;3696].
## [2017-10-04 22:11:48] Aggregated 141020 fragments [1216;3696] to 89063 fragments [1216;1852].
## [2017-10-04 22:11:49] Coerced TopDownSet into an NCBSet object; 69454 fragments [152;1852].</code></pre>
<p>An <code>NCBSet</code> is very similar to a <code>TopDownSet</code> but instead of an <code>FragmentViews</code> the <code>rowViews</code> are an <code>XStringViews</code> for the former. Another difference is that the <code>NCBSet</code> has one row per bond instead one row per fragment. Also the <code>assayData</code> contains no intensity information but a <code>1</code> for an <em>N</em>-terminal, a <code>2</code> for a <em>C</em>-terminal and a <code>3</code> for both fragments.</p>
<p>The <code>NCBSet</code> can be used to select the combination of conditions that provide the best fragment coverage. While computing coverage <code>topdownr</code> awards 1 point for every fragment going from every bond in either <em>N</em> or <em>C</em> directions. This means that bonds covered in both directions increase the score of a condition by 2 points. For the myoglobin fragmentation example we get the following table for the best three conditions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/NCBSet-class.html">bestConditions</a></span>(myoglobinNcb, <span class="dt">n=</span><span class="dv">3</span>)</code></pre></div>
<pre><code>##                                         index fragments
## C0893.10_1.0e+06_1.0e+06_05.00_14_00_1   1065       143
## C1211.70_1.0e+05_0.0e+00_00.00_28_00_05  1253        63
## C0707.30_5.0e+05_5.0e+06_02.50_07_00_1    291        27</code></pre>
</div>
<div id="building-a-fragmentation-map" class="section level2">
<h2 class="hasAnchor">
<a href="#building-a-fragmentation-map" class="anchor"></a>Building a Fragmentation Map</h2>
<p>Fragmentation maps allow visualising the type of fragments produced by fragmentation conditions and their overall distribution along the protein backbone. It also illustrates how the combination of conditions results in a cumulative increase in fragment coverage. Shown below is a fragmentation map for myoglobin <em>m/z</em> 707.3, AGC target <code>1e6</code> and ETD reagent target of <code>1e7</code> for ETD (plotting more conditions is not practical for the vignette):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sel &lt;-
<span class="st">    </span>myoglobin$Mz ==<span class="st"> </span><span class="fl">707.3</span> &amp;
<span class="st">    </span>myoglobin$AgcTarget ==<span class="st"> </span><span class="fl">1e6</span> &amp;
<span class="st">    </span>myoglobin$EtdReagentTarget ==<span class="st"> </span><span class="fl">1e7</span>

myoglobinNcbSub &lt;-<span class="st"> </span>myoglobinNcb[, sel]

<span class="kw"><a href="../reference/NCBSet-class.html">fragmentationMap</a></span>(
    myoglobinNcbSub,
    <span class="dt">nCombinations =</span> <span class="dv">10</span>,
    <span class="dt">labels =</span> <span class="kw">seq_len</span>(<span class="kw">ncol</span>(myoglobinNcbSub))
)</code></pre></div>
<p><img src="analysis_files/figure-html/fragmentationMap-1.png" width="672"></p>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.3 (2017-03-06)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## locale:
##  [1] LC_CTYPE=de_DE.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=de_DE.UTF-8    
##  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=de_DE.UTF-8   
##  [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] BiocStyle_2.2.1     ggplot2_2.2.1       ranger_0.8.0       
##  [4] topdownrdata_0.99.0 topdownr_0.3        Biostrings_2.42.1  
##  [7] XVector_0.14.1      IRanges_2.8.2       S4Vectors_0.12.2   
## [10] BiocGenerics_0.20.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.13          highr_0.6             BiocInstaller_1.24.0 
##  [4] plyr_1.8.4            iterators_1.0.8       ProtGenerics_1.9.1   
##  [7] tools_3.3.3           zlibbioc_1.20.0       MALDIquant_1.16.4    
## [10] digest_0.6.12         evaluate_0.10.1       tibble_1.3.4         
## [13] preprocessCore_1.36.0 gtable_0.2.0          lattice_0.20-34      
## [16] rlang_0.1.2.9000      Matrix_1.2-11         foreach_1.4.3        
## [19] yaml_2.1.14           stringr_1.2.0         knitr_1.17           
## [22] rprojroot_1.2         grid_3.3.3            Biobase_2.34.0       
## [25] impute_1.48.0         XML_3.98-1.9          BiocParallel_1.8.2   
## [28] rmarkdown_1.6         limma_3.30.13         reshape2_1.4.2       
## [31] mzR_2.11.12           magrittr_1.5          backports_1.1.1      
## [34] scales_0.5.0          pcaMethods_1.66.0     codetools_0.2-15     
## [37] htmltools_0.3.6       mzID_1.12.0           MSnbase_2.3.11       
## [40] colorspace_1.3-2      labeling_0.3          affy_1.52.0          
## [43] stringi_1.1.5         doParallel_1.0.11     lazyeval_0.2.0       
## [46] munsell_0.4.3         vsn_3.42.3            affyio_1.44.0</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-SummarizedExperiment">
<p>Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2017. <em>SummarizedExperiment: SummarizedExperiment Container</em>.</p>
</div>
<div id="ref-ranger">
<p>Wright, Marvin N., and Andreas Ziegler. 2017. “ranger: A Fast Implementation of Random Forests for High Dimensional Data in C++ and R.” <em>Journal of Statistical Software</em> 77 (1): 1–17. doi:<a href="https://doi.org/10.18637/jss.v077.i01">10.18637/jss.v077.i01</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#foreword">Foreword</a></li>
      <li><a href="#questions-and-bugs">Questions and bugs</a></li>
      <li>
<a href="#introductionworking-with-topdownr">Introduction/Working with <code>topdownr</code></a><ul class="nav nav-pills nav-stacked">
<li><a href="#importing-files">Importing Files</a></li>
      <li><a href="#the-topdownset-anatomy">The <code>TopDownSet</code> Anatomy</a></li>
      <li><a href="#technical-details">Technical Details</a></li>
      <li><a href="#subsetting-a-topdownset">Subsetting a <code>TopDownSet</code></a></li>
      </ul>
</li>
      <li>
<a href="#fragmentation-data-analysis-of-myoglobin">Fragmentation Data Analysis of Myoglobin</a><ul class="nav nav-pills nav-stacked">
<li><a href="#filter-conditions-on-injection-times">Filter Conditions on Injection Times</a></li>
      <li><a href="#filter-fragments-on-cv">Filter Fragments on CV</a></li>
      <li><a href="#filter-fragments-on-intensity">Filter Fragments on Intensity</a></li>
      <li><a href="#data-aggregation">Data Aggregation</a></li>
      <li><a href="#random-forest">Random Forest</a></li>
      <li><a href="#combining-fragmentation-conditions-to-maximize-coverage">Combining Fragmentation Conditions to Maximize Coverage</a></li>
      <li><a href="#building-a-fragmentation-map">Building a Fragmentation Map</a></li>
      </ul>
</li>
      <li><a href="#session-information">Session Information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Sebastian Gibb, Pavel Shliaha, Ole Nørregaard Jensen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
